jobserver/releases.py

```diff
@@ -0,0 +1,66 @@
+import hashlib
+import os
+import shutil
+from zipfile import ZipFile
+
+from django.conf import settings
+from django.db import transaction
+
+from .models import Release
+
+
+@transaction.atomic
+def handle_release(workspace, backend, backend_user, release_hash, upload):
+    try:
+        # check for duplicates
+        existing_release = Release.objects.get(id=release_hash)
+        return existing_release, False
+    except Release.DoesNotExist:
+        pass
+
+    # Even though id's are globally unique, we partition the filesystem
+    # structure into workspace directories, for 2 reasons:
+    # 1. avoid dumping everything in one big directory.
+    # 2. much easier for human operators to navigate on disk (e.g. for removal
+    #    of any disclosive information).
+    upload_dir = os.path.join(workspace.name, release_hash)
+    actual_dir = settings.RELEASE_STORAGE / upload_dir
+    actual_dir.parent.mkdir(exist_ok=True, parents=True)
+
+    try:
+        # first, ensure we can extract the zip
+        with ZipFile(upload) as archive:
+            archive.extractall(actual_dir)
+
+        calculated_hash = hash_files(actual_dir)
+        if calculated_hash != release_hash:
+            raise Exception(
+                f"provided hash {release_hash} did not match files hash of {calculated_hash}"
+            )
```

Comment:
Consider raising `rest_framework.exceptions.ValidationError` here?

Reply:
arg, yes, I updated the others to ValidationError, but missed this one.

Reply:
I remember why! I was trying to figure out how to return a  412 Pre-condition failed